### ISSUE TITLE
AirGap Install: JWT Token Parsing Issue with Artifactory

### ISSUE THREAD

**Issue by adamcoppard** [replicatedhq/kots#709]
When installing a kots application via `kubectl kots install` in an air-gapped environment we are facing an issue with kots handling the token returned by the Docker registry we are targeting - in this case, Artifactory.

Following the process outlined in the Docker [registry spec](https://docs.docker.com/registry/spec/auth/token/) and logs from Artifactory, we can see that kots responds to the authentication challenge, and receives a response containing the token. However, Artifactory does not respond with a JWT, rather its own token. From the Artifactory logs, we can see `200` responses to the token request from kots, i.e:
```
/artifactory/api/docker/<repo>/v2/token?service=<registry>&scope=repository:kots/testrepo:push
```

From the looks of `TestPushAccess` that responds to the authentication challenge, there is already handling for GCR which responds with a different token format https://github.com/replicatedhq/kots/blob/b0ed907190a1045a122fd966d2676454c8987be4/pkg/docker/registry/auth.go#L118 - but our request continues to attempt to be parsed as a JWT token, before failing - from the looks of things, here https://github.com/replicatedhq/kots/blob/b0ed907190a1045a122fd966d2676454c8987be4/pkg/docker/registry/auth.go#L130

This results in the error as shown below.

![kots-airgap-artifactory](https://user-images.githubusercontent.com/3635400/86048567-a19c2d00-ba48-11ea-818d-aa804b649bce.png)

It is worth noting a `kubectl kots pull` referencing the same Artifactory registry we use in `kubectl kots install` does push the images to the registry.

**Comment by dexhorthy** [comment]
@adamcoppard thanks for filing, we're discussing this week. We likely need to do some testing with `kotsadm` + Artifactory.

Very helpful to know that using `kubectl kots pull` with artifactory does work

**Comment by adamcoppard** [comment]
@dexhorthy - I've been looking at some code changes that would make this work - especially based on the prior art of how `gcr.io` tokens are handled.

The proposal I had in my head was to add an additional conditional for going no further in the case it was Artifactory (as determined by the presence of `/artifactory/api/docker` in the `authURL`) at the same point the `endpoint` string is checked for `gcr.io` and updating the comment to match - purely based on the prior art of the `gcr.io` case.

In terms of the general case though - I wonder if the parsing as a JWT token could be more defensive anyway - as this isn't guaranteed. If that parsing fails, just treat the token as an opaque one until usage time, but if it was a JWT continue with the code as is.

If you'd be open to accepting a PR - let me know.

**Comment by marccampbell** [comment]
@adamcoppard Thanks for the detailed investigation here.

We are open to either solution presented here. I imagine that handling non-JWT tokens as opaque, one time tokens will provide compatibility with registries other than Artifactory, but this is both unknown and doesn't have to be in scope here. 

KOTS currently has some special case handling for certain registries, as you pointed out. Artifactory is a popular registry, and we would not be opposed to a solution that looks for artifactory in the authURL and has special-case handling for that protocol.

**Comment by laverya** [comment]
Would it be possible to copy the `docker login` validation code instead of attempting to rebuild it ourselves?

**Comment by rushabh-harness** [comment]
we have seen this issue happening with nexus repo as well cc @laverya

**Comment by laverya** [comment]
I believe the best resolution here is to treat "this isn't a JWT" as an info-level event, not a fatal error, just like we currently do for `gcr.io`

**Comment by laverya** [comment]
This may be fixed by https://github.com/replicatedhq/kots/pull/749

**Comment by marccampbell** [comment]
@adamcoppard We've released a 1.16.2 that doesn't require the JWT token is validated any longer. If the login response is not a JWT, a message will be logged but the registry settings will save and continue.
